### ISSUE TITLE
Status quo for selecting @TestTemplate or @TestFactory invocations by Unique ID

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#1031]
## Overview

This issue is a prerequisite for #1025.

## Implementation Notes

- Update `JavaElementsResolver.resolveUniqueId(TestDescriptor, List<Segment>)` so that it returns the number of resolved segments instead of a `boolean` and then check that all segments were resolved in `JavaElementsResolver.resolveUniqueId(UniqueId)` in order to determine what information to include in the log message.

## Deliverables

- [x] Test the _status quo_ for selecting `@TestTemplate` and `@TestFactory` invocations by Unique ID.
  - See `DiscoverySelectorResolverTests`.
- [x] Update the warning logged by `JavaElementsResolver.resolveUniqueId()` so that it informs the user that the _Unique ID_ could only be _partially resolved_ and that all test template invocations or dynamic tests will be executed.
  - See _Implementation Notes_.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
FYI: the following test methods in `DiscoverySelectorResolverTests` are in place to verify the status quo.

- `testFactoryMethodResolutionByUniqueId()`
- `testTemplateMethodResolutionByUniqueId()`
- `resolvingDynamicTestByUniqueIdResolvesOnlyUpToParentTestFactory()`
- `resolvingTestTemplateInvocationByUniqueIdResolvesOnlyUpToParentTestTemplate()`

**Comment by sbrannen** [comment]
Log warnings for partial resolution now result in output similar to the following.

```
WARN  ...JavaElementsResolver - Unique ID '[engine:junit-jupiter]/[class:org.junit.jupiter.engine.discovery.DiscoverySelectorResolverTests2$MyTestClass]/[test-factory:dynamicTest()]/[dynamic-test:#1]' could only be partially resolved; failed to resolve the following segments: [Segment [type = 'dynamic-test', value = '#1']]
WARN  ...JavaElementsResolver - Unique ID '[engine:junit-jupiter]/[class:org.junit.jupiter.engine.discovery.DiscoverySelectorResolverTests2$TestClassWithTemplate]/[test-template:testTemplate()]/[test-template-invocation:#1]' could only be partially resolved; failed to resolve the following segments: [Segment [type = 'test-template-invocation', value = '#1']]
```

**Comment by sbrannen** [comment]
Latest, updated version of log warnings with additional information.

```
WARN ...JavaElementsResolver - Unique ID '[engine:junit-jupiter]/[class:org.junit.jupiter.engine.discovery.MyTestClass]/[test-factory:dynamicTest()]/[dynamic-test:#1]' could only be partially resolved. All resolved segments will be executed; however, the following segments could not be resolved: [Segment [type = 'dynamic-test', value = '#1']]

WARN ...JavaElementsResolver - Unique ID '[engine:junit-jupiter]/[bogus:enigma]' could only be partially resolved. All resolved segments will be executed; however, the following segments could not be resolved: [Segment [type = 'bogus', value = 'enigma']]

WARN ...JavaElementsResolver - Unique ID '[engine:junit-jupiter]/[class:org.junit.jupiter.engine.discovery.TestClassWithTemplate]/[test-template:testTemplate()]/[test-template-invocation:#1]' could only be partially resolved. All resolved segments will be executed; however, the following segments could not be resolved: [Segment [type = 'test-template-invocation', value = '#1']]
```

**Comment by marcphilipp** [comment]
👍

**Comment by sbrannen** [comment]
Resolved in `master` in b9813b73b2aa40c3d7e79ecc27948f54f6c66bc1 and 279136529610d9d1954b415345fdfe31fabe36d5.
